Add tag type fix

dombartenope opened the pull request:
# Description

## One Line Summary
This PR fixes an issue where addTag will set an empty string, deleting existing keys, if a number is set as the value. Previously this value was of type 'string', and it will now booleans and numbers by casting the value to a string before sending the request to the OneSignal sever.


## Details

### Motivation
Reports of crashes due to sending unix timestamps as integers, when added as a value to the addTag method. 

On the current major release / latest update, if passing in a unix timestamp as the value argument to the addTags method, you will see an empty string is sent as the value when making a change. This would delete a data tag per other SDK logic that determines an empty value to be the same as a "remove" request.

### Scope

### OPTIONAL - Other

The changes to the RTCOneSignalEventEmitter, are to determine whether the type is an NSNumber or not. The changes made to index.ts were for intellisense (allowing the boolean and number types to be used without an error thrown by the IDE) and to set the value to a string in case the value was a boolean (which would default to 1 or 0 otherwise).
# Testing


## Manual testing

I have tested this on both Android and iOS devices:
![image](https://github.com/OneSignal/react-native-onesignal/assets/56173293/c3ba10f3-0f73-4e04-9476-dfcb274589bb)
![image](https://github.com/OneSignal/react-native-onesignal/assets/56173293/e3e3660e-1fd2-48e7-bf07-7f4025e0048d)


# Affected code checklist

- [ ] Notifications
  - [ ] Display
  - [ ] Open
  - [ ] Push Processing
  - [ ] Confirm Deliveries
- [ ] Outcomes
- [ ] Sessions
- [ ] In-App Messaging
- [ ] REST API requests
- [ ] Public API changes

# Checklist

## Overview

- [x] I have filled out all **REQUIRED** sections above
- [x] PR does one thing
  - If it is hard to explain how any codes changes are related to each other then it most likely needs to be more than one PR
- [ ] Any Public API changes are explained in the PR details and conform to existing APIs

## Testing

- [ ] I have included test coverage for these changes, or explained why they are not needed
- [ ] All automated tests pass, or I explained why that is not possible
- [ ] I have personally tested this on my device, or explained why that is not possible

## Final pass

- [ ] Code is as readable as possible.
  - Simplify with less code, followed by splitting up code into well named functions and variables, followed by adding comments to the code.
- [x] I have reviewed this PR myself, ensuring it meets each checklist item
  - WIP (Work In Progress) is ok, but explain what is still in progress and what you would like feedback on. Start the PR title with "WIP" to indicate this.

<!-- Reviewable:start -->
- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/OneSignal/react-native-onesignal/1696)
<!-- Reviewable:end -->

Commit:
Added logic to handle non string types when sending tags

Commit:
Added logic to handle non string types when sending tags

jennantilla committed:
Update tagging methods to pass string value to native bridge

jennantilla commented:
Hi @dombartenope thank you so much for this PR! 

My observations of this bug and how behavior varied by platform:

`addTag`:
- iOS: replaced int value with empty string (thereby deleting the previous tag value)
- Android: crashed

`addTags`:
- iOS: crashed
- Android: did not add tag with int value but added tag with string value

Since all the SDKs require a string value parameter for tagging methods, I've reverted the `addTag` function signature. If the developer is using Typescript, they should get a warning that they should only pass in a string value. However, we'll ensure that no sneaky non-string values get to the bridge by converting them within the RN methods. 

Let me know if you have any questions or concerns!

nan-li commented on src/index.ts:
```diff
@@ -456,6 +461,14 @@ export namespace OneSignal {
+      Object.keys(tags).forEach(function (key) {
+        // forces values to be string types
+        if (typeof convertedTags[key] !== 'string') {
+          convertedTags[key] = JSON.stringify(convertedTags[key]);
```
Is there a reason this one is using `JSON.stringify` and addTag is using the `String()` constructor above?

nan-li commented on src/index.ts:
```diff
@@ -433,11 +433,16 @@ export namespace OneSignal {
       if (!isNativeModuleLoaded(RNOneSignal)) return;
 
-      if (!key || (!value && value !== '')) {
+      if (!key || !value) {
```
If you want to allow people to pass in `false` as the value, it will not pass this check.

jennantilla replied:
Oops! Updated to check if null or undefined.

jennantilla committed:
Refine tagging methods

- ensure consistency of string conversion methods
- check for null or undefined values instead of falsy checks

jennantilla committed:
Add console warnings for non-string type key values

jennantilla committed:
run prettier
